search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,603 @@
+        return FALSE;
+
+    return g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (self->skeleton),
+                                             connection,
+                                             object_path,
+                                             error);
+}
+
+static void
+yelp_search_provider_app_dbus_unregister (GApplication    *app,
+                                          GDBusConnection *connection,
+                                          const gchar     *object_path)
+{
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    GDBusInterfaceSkeleton *skeleton = G_DBUS_INTERFACE_SKELETON (self->skeleton);
+
+    if (g_dbus_interface_skeleton_has_connection (skeleton, connection))
+        g_dbus_interface_skeleton_unexport_from_connection (skeleton, connection);
+
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->dbus_unregister (app,
+                                                                                  connection,
+                                                                                  object_path);
+}
+
+static void
+yelp_search_provider_app_init (YelpSearchProviderApp *self)
+{
+    self->skeleton = yelp_shell_search_provider2_skeleton_new ();
+
+    g_signal_connect (self->skeleton, "handle-get-initial-result-set",
+                      G_CALLBACK (handle_get_initial_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-subsearch-result-set",
+                      G_CALLBACK (handle_get_subsearch_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                      G_CALLBACK (handle_get_result_metas), self);
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                      G_CALLBACK (handle_activate_result), self);
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                      G_CALLBACK (handle_launch_search), self);
+
+    g_application_set_inactivity_timeout (G_APPLICATION (self), 
+                                          SEARCH_PROVIDER_INACTIVITY_TIMEOUT);
+}
+
+static void
+yelp_search_provider_app_class_init (YelpSearchProviderAppClass *klass)
+{
+    GApplicationClass *aclass = G_APPLICATION_CLASS (klass);
+    GObjectClass *oclass = G_OBJECT_CLASS (klass);
+
+    aclass->startup = search_provider_app_startup;
+    aclass->dbus_register = yelp_search_provider_app_dbus_register;
+    aclass->dbus_unregister = yelp_search_provider_app_dbus_unregister;
+
+    oclass->dispose = search_provider_app_dispose;
+}
+
+static GApplication *
+yelp_search_provider_app_new (void)
+{
+    return g_object_new (YELP_TYPE_SEARCH_PROVIDER_APP,
+                         "application-id", "org.gnome.Yelp.SearchProvider",
+                         "flags", G_APPLICATION_IS_SERVICE,
+                         NULL);
+}
+
+int
+main (int    argc,
+      char **argv)
+{
+    GApplication *app;
+    gint res;
+    gchar *env_var = "YELP_SKIP_GUI";
```

Comment:
So I modified the code a bit - basically commented out the g_setenv thing and added `gtk_init (&argc, &argv);`. I'm pasting two backtraces here, one with `G_DEBUG=fatal-criticals` and one without. But before just some description what happens and why we decided to resolve the crash with the (admittedly hacky) `YELP_SKIP_GUI` env var: the crash is related to GtkTextView. Doing `git grep -nie GtkTextView` in libyelp gives no results, but `git grep -nie gtk_text_view` gives us just one: [`yelp-settings.c:1019`](https://github.com/endlessm/yelp/blob/master/libyelp/yelp-settings.c#L1012)`:    tmpview = gtk_text_view_new ();`. Basically there is the `gtk_theme_changed` callback function that gathers some colors from the widgets to put them inside YelpSettings and this callback is connected to the "notify::gtk-theme-name" GtkSettings' signal and is also executed directly in `yelp_settings_set_property` for the `PROP_GTK_SETTINGS` property. Since we are absolutely not interested in colors in the search provider, we just added a possibility to skip all this code with an env var. I admit that this is a workaround here instead of a proper solution, but I have a feeling that getting colors that way is just plain wrong and probably some CSS should be used instead, so probably fixing it properly is a too big task for now.

Incidentally, you maybe will want to see the `case PROP_GTK_ICON_THEME` in the property setter to see what shenanigans yelp is doing with icons (or icon theme), which maybe explains our problems with icons in the search provider…

So, the backtraces:

With `G_DEBUG=fatal-criticals` (message is `(yelp-search-provider:3243): Gtk-CRITICAL **: gtk_text_attributes_ref: assertion 'values != NULL' failed`):
```
#0  0x00007ffff0317a31 in _g_log_abort (breakpoint=breakpoint@entry=1) at gmessages.c:509
#1  0x00007ffff0318e3d in g_logv (log_domain=0x7ffff3338a38 "Gtk", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7fffffffd600) at gmessages.c:1318
#2  0x00007ffff0318f9f in g_log (log_domain=log_domain@entry=0x7ffff3338a38 "Gtk", log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7ffff0363c85 "%s: assertion '%s' failed") at gmessages.c:1359
#3  0x00007ffff0319419 in g_return_if_fail_warning (log_domain=log_domain@entry=0x7ffff3338a38 "Gtk", pretty_function=pretty_function@entry=0x7ffff339fd30 <__func__.53067> "gtk_text_attributes_ref", expression=expression@entry=0x7ffff335ced0 "values != NULL") at gmessages.c:2536
#4  0x00007ffff3253ab3 in gtk_text_attributes_ref (values=0x0) at gtktextattributes.c:193
#5  0x00007ffff3278860 in get_style (layout=0x7fffd0007400 [GtkTextLayout], tags=0x0) at gtktextlayout.c:1223
#6  0x00007ffff327b0fd in gtk_text_layout_get_line_display (layout=layout@entry=0x7fffd0007400 [GtkTextLayout], line=line@entry=0x7fffc4006260, size_only=size_only@entry=1)
    at gtktextlayout.c:2345
#7  0x00007ffff327bb62 in gtk_text_layout_real_wrap (layout=0x7fffd0007400 [GtkTextLayout], line=0x7fffc4006260, line_data=0x7fffc40014a0) at gtktextlayout.c:1182
#8  0x00007ffff3259ce5 in _gtk_text_btree_validate_line (tree=<optimized out>, line=line@entry=0x7fffc4006260, view_id=view_id@entry=0x7fffd0007400) at gtktextbtree.c:5454
#9  0x00007ffff3279796 in gtk_text_layout_validate_yrange (layout=0x7fffd0007400 [GtkTextLayout], anchor=anchor@entry=0x7fffffffd9e0, y0=y0@entry=0, y1=200)
    at gtktextlayout.c:1091
#10 0x00007ffff328b640 in gtk_text_view_validate_onscreen (text_view=0x76ee70 [GtkTextView]) at gtktextview.c:4394
#11 0x00007ffff328c1b3 in gtk_text_view_flush_first_validate (text_view=0x76ee70 [GtkTextView]) at gtktextview.c:4452
#12 0x00007ffff328c5e9 in first_validate_callback (data=<optimized out>) at gtktextview.c:4471
#13 0x00007ffff2ca4e78 in gdk_threads_dispatch (data=data@entry=0x7fffc4001460) at gdk.c:743
#14 0x00007ffff030e8e7 in g_idle_dispatch (source=0x652ce0, callback=0x7ffff2ca4e50 <gdk_threads_dispatch>, user_data=0x7fffc4001460) at gmain.c:5545
#15 0x00007ffff0311e52 in g_main_dispatch (context=0x631ed0) at gmain.c:3203
#16 0x00007ffff0311e52 in g_main_context_dispatch (context=context@entry=0x631ed0) at gmain.c:3856
#17 0x00007ffff03121d0 in g_main_context_iterate (context=context@entry=0x631ed0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at gmain.c:3929
#18 0x00007ffff031227c in g_main_context_iteration (context=context@entry=0x631ed0, may_block=may_block@entry=1) at gmain.c:3990
#19 0x00007ffff18bdb9d in g_application_run (application=0x7fffd00036b0 [YelpSearchProviderApp], argc=1, argv=0x7fffffffdd48) at gapplication.c:2381
#20 0x0000000000407317 in main (argc=1, argv=0x7fffffffdd48) at yelp-search-provider.c:587
```

And one without (so we have SIGSEGV):
```
#0  0x00007ffff3254858 in gtk_text_btree_node_ensure_data (view_id=view_id@entry=0x64c510, node=<optimized out>) at gtktextbtree.c:5555
#1  0x00007ffff3257367 in gtk_text_btree_node_check_valid (node=node@entry=0x683820, view_id=view_id@entry=0x64c510) at gtktextbtree.c:5368
#2  0x00007ffff3259cfb in gtk_text_btree_node_check_valid_upward (view_id=0x64c510, node=0x683820) at gtktextbtree.c:5388
#3  0x00007ffff3259cfb in _gtk_text_btree_validate_line (tree=<optimized out>, line=line@entry=0x7fffc401c060, view_id=view_id@entry=0x64c510) at gtktextbtree.c:5456
#4  0x00007ffff3279796 in gtk_text_layout_validate_yrange (layout=0x64c510, anchor=anchor@entry=0x7fffffffda10, y0=y0@entry=0, y1=200) at gtktextlayout.c:1091
#5  0x00007ffff328b640 in gtk_text_view_validate_onscreen (text_view=0x75aec0) at gtktextview.c:4394
#6  0x00007ffff328c1b3 in gtk_text_view_flush_first_validate (text_view=0x75aec0) at gtktextview.c:4452
#7  0x00007ffff328c5e9 in first_validate_callback (data=<optimized out>) at gtktextview.c:4471
#8  0x00007ffff2ca4e78 in gdk_threads_dispatch (data=data@entry=0x646e80) at gdk.c:743
#9  0x00007ffff030e8e7 in g_idle_dispatch (source=0x77b300, callback=0x7ffff2ca4e50 <gdk_threads_dispatch>, user_data=0x646e80) at gmain.c:5545
#10 0x00007ffff0311e52 in g_main_dispatch (context=0x631ed0) at gmain.c:3203
#11 0x00007ffff0311e52 in g_main_context_dispatch (context=context@entry=0x631ed0) at gmain.c:3856
#12 0x00007ffff03121d0 in g_main_context_iterate (context=context@entry=0x631ed0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at gmain.c:3929
#13 0x00007ffff031227c in g_main_context_iteration (context=context@entry=0x631ed0, may_block=may_block@entry=1) at gmain.c:3990
#14 0x00007ffff18bdb9d in g_application_run (application=0x7fffd00036b0 [YelpSearchProviderApp], argc=1, argv=0x7fffffffdd78) at gapplication.c:2381
#15 0x0000000000407317 in main (argc=1, argv=0x7fffffffdd78) at yelp-search-provider.c:587
```